src/main/java/com/team766/controllers/ExtendableCanSparkMaxSmartMotionPIDController.java

```diff
@@ -0,0 +1,283 @@
+package com.team766.controllers;
+
+import com.team766.config.ConfigFileReader;
+import com.team766.hal.RobotProvider;
+import com.team766.library.SetValueProvider;
+import com.team766.library.SettableValueProvider;
+import com.team766.library.ValueProvider;
+import com.team766.logging.Logger;
+import com.team766.logging.LoggerExceptionUtils;
+import edu.wpi.first.math.MathUtil;
+import com.revrobotics.SparkMaxAbsoluteEncoder;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.ControlType;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.revrobotics.SparkMaxAbsoluteEncoder.Type;
+import java.io.IOError;
+import javax.swing.text.DefaultStyledDocument.ElementSpec;
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.revrobotics.CANSparkMax;
+import com.team766.framework.Mechanism;
+import com.team766.hal.MotorController;
+import com.team766.library.RateLimiter;
+import com.team766.controllers.PIDSTATE.*;
+import com.team766.framework.Exceptions.*;
+
+
+public abstract class ExtendableCanSparkMaxSmartMotionPIDController {
+		/*
+		 * These are the hardware variables for the hardware we have on the robot.
+		 * It includes a CanSparkMax, MotorController of any type, and a SparkMaxAbsoluteEncoder.
+		 * @author Max Spier - 9/9/2023
+		 */
+		private MotorController mc;
+		private CANSparkMax csm;
+		private SparkMaxAbsoluteEncoder abs;
+	
+		/*
+		 * These are PID related variables that we will store locally to run PIDs
+		 * We also create an object for the PIDController
+		 * @author Max Spier - 9/9/2023
+		 */
+		private SparkMaxPIDController pid;
+		private static double deadzone = 0;
+		private static double setPointPosition = 0;
+		private static double comboOfTimesInsideDeadzone = 0;
+		private static double minPos = 0;
+		private static double maxPos = 0;
+	
+		/*
+		 * These are some variables that we can use so that we can reset the motor controller
+		 * We use this so that we can get an accurate ratio for the hall encoder units to absolute encoder units.
+		 * @author Max Spier - 9/9/2023
+		 */
+		private double encoderUnitsPerOneAbsolute = 0;
```

Comment:
if some of these variables must never be zero, force them to be specified within the constructor
otherwise, provide a reasonable default

Reply:
hmm. I don't think that will work -- every encoder pattern is different. Instead, ill catch the devide by 0 error and throw a new exception